### ISSUE TITLE
Expose SMS sending capabilities in android wrapper

### ISSUE THREAD

**Issue by SCdF** [medic/medic#4811]
We need to allow for the web app to send SMS messages. We can achieve this (on phones) by exposing an API endpoint in the JS bridge that lets you pass a phone number and a message, which the android wrapper will then send.

We do not need to read SMS messages or do anything more complicated at this time.

Things to think about:
 - Do we want to just blind-fire this request, or have it call a callback when it succeeds (or fails)?
 - Do we need some kind of queuing on the android wrapper side?

**Comment by alxndrsn** [comment]
> Do we want to just blind-fire this request, or have it call a callback when it succeeds (or fails)?

I've added support for a callback to the `AndroidApi` angular service when sending and delivery reports come in.  Actually supporting these in the webapp UI in a useful way would probably be much more complicated.

> Do we need some kind of queuing on the android wrapper side?

I'm not sure what the motivation would be for this; have you got any more thoughts on it?

Thoughts on UI:

SENT/FAILED status is likely to come back within a few seconds.  DELIVERED/FAILED could theoretically take hours or days, but that would normally depend on the receiving party, and hopefully our gateway will always be on and within reception.

Perhaps we could have the webapp UI block while the SMS is initially sent, and a snackbar message detailing delivery success or failure of the most recent submission.

**Comment by SCdF** [comment]
>> Do we need some kind of queuing on the android wrapper side?

>I'm not sure what the motivation would be for this; have you got any more thoughts on it?

This is my ignorance of how SMS sending works really. If you don't think it's required then that's cool, I just wasn't sure if the OS managed having to retry etc or we do

> Perhaps we could have the webapp UI block while the SMS is initially sent

How would you feel if we just did it async and presumed it worked? There is definitely going to be a non-technology-based backup system in place (since SMS can just be arbitrarily delayed etc). I think if we show progress and potential failures we also need a UI for dealing with retries etc. For now I'm OK with it just happening in the background, but I'm interested in your expert opinion.

**Comment by alxndrsn** [comment]
> This is my ignorance of how SMS sending works really. If you don't think it's required then that's cool, I just wasn't sure if the OS managed having to retry etc or we do

We'd have to do it, I think.  I expect that's how it works for user-generated text SMS in the default SMS application too.

> How would you feel if we just did it async and presumed it worked?

Definitely an option.  I think there are different levels of async, and the current implementation probably blocks the UI... which means we'll need it async at least on that level.

**Comment by SCdF** [comment]
@alxndrsn we merged the medic-webapp side, but we haven't merged the android side yet: https://github.com/medic/medic-android/pull/62

Do you want to merge that, or hold off (and revert the merge of the webapp callback hook so we don't leave confusing code around) because it's been deprioritised?

**Comment by newtewt** [comment]
@SCdF  can you summarize the changes that happened for this? Also, I think this may have been merged into master already as I can't find a linked PR that's open. Any thoughts?

**Comment by SCdF** [comment]
Yeah it was already merged. I don't think there is anything specific to AT here, since we're not actually engaging in using it in this ticket, it was more of a backend-y thing.

**Comment by newtewt** [comment]
Sounds good. Closing then.
